File: cpp/simpleGRpcServer.cpp
#include <iostream>
#include <grpcpp/server_builder.h>
#include <simple_proto.grpc.pb.h>

using grpc::Status;
using grpc::ServerContext;
using grpc::ServerBuilder;
using grpc::Server;

class SimpleGRpcServerImpl final : public SimpleGRpc::Service
{
public:
    Status Echo(ServerContext *context, const SimpleData *request, SimpleData *response) override {
        response->set_data(request->data());
        return Status::OK;
    }

    Status Add(ServerContext *context, const SimpleTwoData *request, SimpleResult *response) override {
        response->set_result(request->input1() + request->input2());
        return Status::OK;
    }
};

void runServer() {
    std::string serverAddress("0.0.0.0:50051");
    SimpleGRpcServerImpl service;

    ServerBuilder builder;
    builder.AddListeningPort(serverAddress, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "Server listening on " << serverAddress << std::endl;
    server->Wait();
}
int main() {
    //std::cout << "Hello, World!" << std::endl;
    runServer();
    return 0;
}

File: python/simpleGRpcServer.py
from concurrent import futures
import grpc
import simple_proto_pb2
import simple_proto_pb2_grpc

import logging


class SimpleGRpcServicer(simple_proto_pb2_grpc.SimpleGRpcServicer):
    def Echo(self, request, context):
        return simple_proto_pb2.SimpleData(data=request.data)

    def Add(self, request, context):
        result = request.input1 + request.input2
        return simple_proto_pb2.SimpleResult(result=result)


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    simple_proto_pb2_grpc.add_SimpleGRpcServicer_to_server(
        SimpleGRpcServicer(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    server.wait_for_termination()


if __name__ == '__main__':
    logging.basicConfig()
    serve()
